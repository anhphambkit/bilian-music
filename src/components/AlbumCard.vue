<template>
	<div class="album-card">
		<slot>
			<div class="album-card__order">
				<div class="album-card__info order">
					{{ order }}
				</div>
			</div>
			<div class="album-card__info-wrapper">
				<div class="album-card__main">
					<div class="album-card__info album">
						<div class="album__image-wrapper">
							<b-img-lazy
								rounded
								class="album__img"
								:alt="albumName"
								:src="albumImage"
							/>
						</div>
						<div class="album__name" v-b-tooltip.hover :title="albumName">
							{{ albumName }}
						</div>
					</div>
				</div>
				<div class="album-card__sub">
					<div class="album-card__info author">
						<div class="album-card__info--value">
							<div class="author__image-wrapper">
								<b-img-lazy
									rounded="circle"
									class="author__img"
									:alt="artistName"
									:src="artistImage"
								/>
							</div>
							<div class="author__name" v-b-tooltip.hover :title="artistName">
								{{ artistName }}
							</div>
						</div>
						<div class="album-card__info--name">Artist</div>
					</div>
					<div class="album-card__info total-tracks">
						<div class="album-card__info--value">
							{{ trackCount }}
						</div>
						<div class="album-card__info--name">Total tracks</div>
					</div>
					<div class="album-card__info total-disc">
						<div class="album-card__info--value">
							{{ discCount }}
						</div>
						<div class="album-card__info--name">Total Discs</div>
					</div>
					<div class="album-card__info released">
						<div class="album-card__info--value">
							{{ released }}
						</div>
						<div class="album-card__info--name">Released</div>
					</div>
				</div>
			</div>
		</slot>
	</div>
</template>
<script>
export default {
	name: "AlbumCard",
	props: {
		order: {
			type: Number,
			default: 1,
		},
		discCount: {
			type: Number,
			default: 0,
		},
		trackCount: {
			type: Number,
			default: 0,
		},
		released: {
			type: String,
			required: true,
		},
		albumImage: {
			type: String,
			required: true,
		},
		albumName: {
			type: String,
			required: true,
		},
		artistImage: {
			type: String,
			required: true,
		},
		artistName: {
			type: String,
			required: true,
		},
	},
};
</script>
<style lang="scss" scoped>
.album-card {
	display: flex;
	background-color: #262626;
	padding: 10px 0;
	border-radius: 5px;
	color: white;
	align-items: center;
	justify-content: space-around;
	transition: all 0.3s ease;
	flex-wrap: wrap;
	&:hover {
		box-shadow: 0px 2px 10px rgb(0, 0, 0);
	}
	&__order {
		width: 5%;
		@media screen and (max-width: 992px) {
			width: 10%;
		}
	}
	&__info-wrapper {
		width: 95%;
		display: flex;
		align-items: center;
		justify-content: center;
		flex-wrap: wrap;

		@media screen and (max-width: 992px) {
			width: 90%;
		}
	}
	&__main {
		display: flex;
		align-items: center;
		width: 50%;
		margin-bottom: 10px;
		@media screen and (max-width: 992px) {
			width: 100%;
		}
	}
	&__sub {
		display: flex;
		align-items: center;
		width: 50%;
		@media screen and (max-width: 992px) {
			width: 100%;
		}
	}
	&__info {
		display: flex;
		align-items: center;
		flex-direction: column;
		width: 20%;
		padding: 0 5px;
		&--value {
			display: flex;
			align-items: center;
			justify-content: center;
			font-size: 14px;
			@media screen and (max-width: 992px) {
				font-size: 10px;
			}
		}
		&--name {
			font-size: 10px;
			text-align: center;
			color: #9c9ca2;
			@media screen and (max-width: 992px) {
				font-size: 8px;
			}
		}
		&.order {
			font-size: 30px;
			font-weight: bold;
			padding-left: 10px;
			width: 100%;
			@media screen and (max-width: 992px) {
				font-size: 20px;
			}
		}
		&.album {
			flex-direction: row;
			width: 90%;
			.album__image-wrapper {
				margin-right: 10px;
				cursor: pointer;
			}
			.album__img {
				width: 100px;
				height: 100px;
				@media screen and (max-width: 992px) {
					width: 40px;
					height: 40px;
				}
			}
			.album__name {
				color: #f47e00;
				cursor: pointer;
			}
			@media screen and (max-width: 992px) {
				.album__name {
					font-size: 12px;
				}
			}
		}
		&.author {
			width: 40%;
			.album-card__info--value {
				justify-content: flex-start;
				width: 100%;
			}
			.author__image-wrapper {
				margin-right: 10px;
				cursor: pointer;
				@media screen and (max-width: 992px) {
					margin-right: 5px;
				}
			}
			.author__img {
				width: 30px;
				height: 30px;
				@media screen and (max-width: 992px) {
					width: 20px;
					height: 20px;
				}
			}
			.album-card__info--name {
				text-align: left;
				width: 100%;
			}
			.author__name {
				cursor: pointer;
			}
		}
		&.released {
			padding-right: 10px;
		}
	}
}
</style>