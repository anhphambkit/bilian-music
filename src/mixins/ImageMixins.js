import DEFAULT_THUMB from "@/assets/media/default.jpg";
export default {
    name: 'ImageMixins',
    mixins: [],
    props: {},
    data() {
      return {}
    },
    computed: {},
    watch: {},
    created() {},
    mounted() {},
    methods: {
        $_imageMixins_loadImageSrc(src, altImage = DEFAULT_THUMB) {
            return src ? src : altImage;
        },
        $_imageMixins_loadImageError(event) {
            let img = event.srcElement;
            img.src = DEFAULT_THUMB;
            event.target.src = DEFAULT_THUMB;
            img.onerror = null; 
        }
    },
}