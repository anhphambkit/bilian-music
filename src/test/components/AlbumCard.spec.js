import { createLocalVue, shallowMount } from '@vue/test-utils'
import AlbumCard from '@/components/AlbumCard'
import ImageMixins from "@/mixins/ImageMixins";

// install plugins as normal
import BootstrapVue from "bootstrap-vue";
function createConfig(overrides) {
    process.env.BOOTSTRAP_VUE_NO_WARN = true
    const albumItem = { 
        released: "2020-10-23T00:00:00.000Z",
        albumId: "alb.517280316",
        albumImage: "https://api.napster.com/imageserver/v2/albums/alb.517280316/images/200x200.jpg",
        albumName: "Year Zero",
        artistImage: "https://api.napster.com/imageserver/v2/artists/art.122743427/images/70x70.jpg",
        artistName: "Tchami",
        artistId: "art.122743427",
        discCount: 2,
    }
    const localVue = createLocalVue()
    localVue.mixin(ImageMixins)
    localVue.use(BootstrapVue)
    const mocks = {
        $router: {
          push: jest.fn()
        },
    }
    const propsData = albumItem
    return Object.assign({ mocks, propsData, localVue }, overrides)
}
  
describe('AlbumCard.vue', () => {
    test('TEST CASE 1: Render AlbumCard HTML', () => {
        const wrapper  = shallowMount(AlbumCard, createConfig())
        expect(wrapper).toMatchSnapshot()
    })

    test('TEST CASE 2: redirect to AlbumDetail page when clicking on image album', () => {
        const config = createConfig()
        const wrapper  = shallowMount(AlbumCard, config)
        wrapper.find('.album__image-wrapper').trigger('click')
        const spy = jest.spyOn(config.mocks.$router, 'push')
        expect(spy).toHaveBeenCalledWith(`/album/${config.propsData.albumId}`)
    })

    test('TEST CASE 3: redirect to AlbumDetail page when clicking on album name', () => {
        const config = createConfig()
        const wrapper  = shallowMount(AlbumCard, config)
        wrapper.find('.album__name').trigger('click')
        const spy = jest.spyOn(config.mocks.$router, 'push')
        expect(spy).toHaveBeenCalledWith(`/album/${config.propsData.albumId}`)
    })

    test('TEST CASE 4: redirect to ArtistDetail page when clicking on author name and author is available(it\'s not various artists)', () => {
        const config = createConfig()
        const wrapper  = shallowMount(AlbumCard, config)
        wrapper.find('.album-card__info--value.author-detail').trigger('click')
        const spy = jest.spyOn(config.mocks.$router, 'push')
        expect(spy).toHaveBeenCalledWith(`/artist/${config.propsData.artistId}`)
    })
})