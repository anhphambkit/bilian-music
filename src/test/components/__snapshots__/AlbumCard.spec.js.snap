// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlbumCard.vue TEST CASE 1: Render AlbumCard HTML 1`] = `
VueWrapper {
  "_emitted": Object {
    "hook:mounted": Array [
      Array [],
    ],
  },
  "_emittedByOrder": Array [
    Object {
      "args": Array [],
      "name": "hook:mounted",
    },
  ],
  "isFunctionalComponent": undefined,
}
`;
